File: code-example-tests/javascript/driver/utils/sampleDataChecker.js
/**
 * Utility for checking MongoDB sample data availability and conditionally skipping tests.
 *
 * This module provides functions to automatically skip tests when sample data is missing,
 * with clear feedback about what's missing and how to fix it.
 *
 * @fileoverview Sample data detection and test skipping utilities for Jest
 */

import { MongoClient } from 'mongodb';

/**
 * Standard MongoDB sample database names and their expected collections
 */
const SAMPLE_DATABASES = {
  sample_mflix: ['movies', 'theaters', 'users', 'comments', 'sessions'],
  sample_restaurants: ['restaurants', 'neighborhoods'],
  sample_training: [
    'posts',
    'companies',
    'inspections',
    'routes',
    'trips',
    'grades',
    'zips',
  ],
  sample_analytics: ['customers', 'accounts', 'transactions'],
  sample_airbnb: ['listingsAndReviews'],
  sample_geospatial: ['shipwrecks'],
  sample_guides: ['planets', 'comets'],
  sample_stores: ['sales'],
  sample_supplies: ['sales'],
  sample_weatherdata: ['data'],
};

/**
 * Cache for sample data availability to avoid repeated database queries
 * @type {Map<string, boolean>}
 */
const sampleDataCache = new Map();

/**
 * Global flag to track if we've shown sample data availability summary
 */
let hasShownSummary = false;

/**
 * Shows a helpful summary of sample data availability (only once per test run)
 * Returns immediately if already shown to avoid duplicate logging
 */
async function showSampleDataSummary() {
  if (hasShownSummary) return;
  hasShownSummary = true;

  try {
    const availableDatabases = await getAvailableSampleDatabases();

    if (availableDatabases.length === 0) {
      console.log('\n📊 Sample Data Status: No MongoDB sample databases found');
      console.log('   Some tests may be skipped. To load sample data:');
      console.log(
        '   • Atlas: https://www.mongodb.com/docs/atlas/sample-data/'
      );
      console.log('   • Local: Use mongorestore with sample data archive\n');
    } else {
      console.log(
        `\n📊 Sample Data Status: ${availableDatabases.length} database(s) available`
      );
      console.log(`   Found: ${availableDatabases.join(', ')}\n`);
    }
  } catch (error) {
    // Silently fail - we don't want to break test runs over this
    console.warn('Could not determine sample data status:', error.message);
  }
}

/**
 * Checks if a specific sample database and its expected collections exist
 * @param {string} databaseName - The sample database name to check
 * @param {string[]} [requiredCollections] - Optional array of specific collections to verify
 * @returns {Promise<boolean>} True if the sample database and collections exist
 *
 * @example
 * // Check if sample_mflix database exists with default collections
 * const hasMovieData = await checkSampleDataAvailable('sample_mflix');
 *
 * // Check if specific collections exist
 * const hasCustomData = await checkSampleDataAvailable('sample_mflix', ['movies', 'theaters']);
 */
export async function checkSampleDataAvailable(
  databaseName,
  requiredCollections = null
) {
  const cacheKey = `${databaseName}:${JSON.stringify(requiredCollections)}`;

  if (sampleDataCache.has(cacheKey)) {
    return sampleDataCache.get(cacheKey);
  }

  try {
    const uri = process.env.CONNECTION_STRING;
    if (!uri) {
      console.warn('CONNECTION_STRING environment variable not set');
      sampleDataCache.set(cacheKey, false);
      return false;
    }

    // Configure client with short timeouts for testing
    const client = new MongoClient(uri, {
      serverSelectionTimeoutMS: 2000, // 2 second timeout for server selection
      connectTimeoutMS: 2000, // 2 second timeout for initial connection
      socketTimeoutMS: 2000, // 2 second timeout for socket operations
    });
    await client.connect();

    try {
      // Check if database exists
      const adminDb = client.db().admin();
      const databases = await adminDb.listDatabases();
      const dbExists = databases.databases.some(
        (db) => db.name === databaseName
      );

      if (!dbExists) {
        sampleDataCache.set(cacheKey, false);
        return false;
      }

      // Check collections if specified
      const collectionsToCheck =
        requiredCollections || SAMPLE_DATABASES[databaseName] || [];
      if (collectionsToCheck.length > 0) {
        const db = client.db(databaseName);
        const collections = await db.listCollections().toArray();
        const existingCollections = collections.map((col) => col.name);

        const missingCollections = collectionsToCheck.filter(
          (col) => !existingCollections.includes(col)
        );

        if (missingCollections.length > 0) {
          sampleDataCache.set(cacheKey, false);
          return false;
        }
      }

      sampleDataCache.set(cacheKey, true);
      return true;
    } finally {
      await client.close();
    }
  } catch (error) {
    console.warn(
      `Error checking sample data availability for ${databaseName}:`,
      error.message
    );
    sampleDataCache.set(cacheKey, false);
    return false;
  }
}

/**
 * Checks if any of the standard MongoDB sample databases are available
 *
 * @returns {Promise<string[]>} Array of available sample database names
 *
 * @example
 * const availableDbs = await getAvailableSampleDatabases();
 * console.log('Available sample databases:', availableDbs);
 */
export async function getAvailableSampleDatabases() {
  const results = await Promise.all(
    Object.keys(SAMPLE_DATABASES).map(async (dbName) => {
      const isAvailable = await checkSampleDataAvailable(dbName);
      return isAvailable ? dbName : null;
    })
  );

  return results.filter((db) => db !== null);
}

/**
 * Creates a Jest test wrapper that skips the test if required sample data is not available
 * Automatically shows sample data summary on first use
 *
 * @param {string} description - Test description
 * @param {Function} testFn - Test function to execute
 * @param {string|string[]} requiredDatabases - Sample database name(s) required for the test
 * @param {Object} [options] - Additional options
 * @param {Object} [options.collections] - Object mapping database names to required collections
 * @returns {void}
 *
 * @example
 * // Test that requires sample_mflix database
 * describeWithSampleData('Movie aggregation tests', () => {
 *   it('should filter movies by genre', async () => {
 *     // Test implementation
 *   });
 * }, 'sample_mflix');
 *
 * // Test that requires multiple databases
 * describeWithSampleData('Cross-database tests', () => {
 *   it('should join movie and restaurant data', async () => {
 *     // Test implementation
 *   });
 * }, ['sample_mflix', 'sample_restaurants']);
 */
export function describeWithSampleData(
  description,
  testFn,
  requiredDatabases,
  options = {}
) {
  const databases = Array.isArray(requiredDatabases)
    ? requiredDatabases
    : [requiredDatabases];

  describe(description, () => {
    let sampleDataAvailable = false;
    let missingDatabases = [];

    beforeAll(async () => {
      // Show sample data summary (only once per test run)
      try {
        await showSampleDataSummary();
      } catch (error) {
        // Ignore summary errors to prevent test failures
      }

      const availabilityChecks = await Promise.all(
        databases.map(async (dbName) => {
          const requiredCollections = options.collections?.[dbName];
          const isAvailable = await checkSampleDataAvailable(
            dbName,
            requiredCollections
          );
          return { dbName, isAvailable };
        })
      );

      missingDatabases = availabilityChecks
        .filter((check) => !check.isAvailable)
        .map((check) => check.dbName);

      sampleDataAvailable = missingDatabases.length === 0;

      if (!sampleDataAvailable) {
        console.warn(
          `\n⚠️  Skipping "${description}" - Missing: ${missingDatabases.join(', ')}`
        );
      }
    });

    if (typeof testFn === 'function') {
      // Call the test definition function, but Jest will skip if needed
      const originalIt = global.it;
      const originalTest = global.test;

      // Override it/test to conditionally skip
      global.it = global.test = (testDescription, testFunc, timeout) => {
        if (!sampleDataAvailable) {
          return originalIt.skip(testDescription, testFunc, timeout);
        }
        return originalIt(testDescription, testFunc, timeout);
      };

      try {
        testFn();
      } finally {
        // Restore original it/test functions
        global.it = originalIt;
        global.test = originalTest;
      }
    }
  });
}

/**
 * Creates a Jest test case that skips if required sample data is not available
 * Automatically shows sample data summary on first use
 *
 * @param {string} description - Test description
 * @param {Function} testFn - Test function to execute
 * @param {string|string[]} requiredDatabases - Sample database name(s) required for the test
 * @param {Object} [options] - Additional options
 * @param {Object} [options.collections] - Object mapping database names to required collections
 * @param {number} [options.timeout] - Test timeout in milliseconds
 * @returns {void}
 *
 * @example
 * describe('Movie tests', () => {
 *   itWithSampleData('should filter movies by year', async () => {
 *     // Test implementation
 *   }, 'sample_mflix');
 *
 *   itWithSampleData('should find restaurants near theaters', async () => {
 *     // Test implementation
 *   }, ['sample_mflix', 'sample_restaurants']);
 * });
 */
export function itWithSampleData(
  description,
  testFn,
  requiredDatabases,
  options = {}
) {
  const databases = Array.isArray(requiredDatabases)
    ? requiredDatabases
    : [requiredDatabases];

  it(
    description,
    async () => {
      // Show sample data summary (only once per test run)
      // Use setImmediate to avoid hanging after test completion
      const summaryPromise = showSampleDataSummary().catch(() => {
        // Ignore summary errors to prevent test failures
      });
      
      const availabilityChecks = await Promise.all(
        databases.map(async (dbName) => {
          const requiredCollections = options.collections?.[dbName];
          const isAvailable = await checkSampleDataAvailable(
            dbName,
            requiredCollections
          );
          return { dbName, isAvailable };
        })
      );

      const missingDatabases = availabilityChecks
        .filter((check) => !check.isAvailable)
        .map((check) => check.dbName);

      if (missingDatabases.length > 0) {
        console.warn(
          `\n⚠️  Skipping "${description}" - Missing: ${missingDatabases.join(', ')}`
        );
        return; // Skip the test
      }

      // Run the test if sample data is available
      await testFn();
      
      // Ensure summary is completed (but don't wait if it hangs)
      await Promise.race([summaryPromise, new Promise(resolve => setTimeout(resolve, 100))]);
    },
    options.timeout
  );
}

/**
 * Clears the sample data availability cache
 * Useful for testing or when sample data availability may have changed
 *
 * @example
 * // Clear cache after loading new sample data
 * clearSampleDataCache();
 */
export function clearSampleDataCache() {
  sampleDataCache.clear();
}

File: code-example-tests/java/utilities/sample-data/src/main/java/sampledatautil/SampleDataChecker.java
package sampledatautil;

import com.mongodb.ConnectionString;
import com.mongodb.MongoClientSettings;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

/**
 * Utility for checking MongoDB sample data availability and conditionally skipping tests.
 *
 * This class provides methods to automatically skip tests when sample data is missing,
 * with clear feedback about what's missing and how to fix it.
 */
public class SampleDataChecker {
    private static final String DEFAULT_CONN_STRING_ENV = "CONNECTION_STRING";

    /**
     * Standard MongoDB sample database names and their expected collections
     */
    public static final Map<String, List<String>> STANDARD_SAMPLE_DATABASES = Map.of(
        "sample_mflix", List.of("movies", "theaters", "users", "comments", "sessions"),
        "sample_restaurants", List.of("restaurants", "neighborhoods"),
        "sample_training", List.of("posts", "companies", "inspections", "routes", "trips", "grades", "zips"),
        "sample_analytics", List.of("customers", "accounts", "transactions"),
        "sample_airbnb", List.of("listingsAndReviews"),
        "sample_geospatial", List.of("shipwrecks"),
        "sample_guides", List.of("planets", "comets"),
        "sample_stores", List.of("sales"),
        "sample_supplies", List.of("sales"),
        "sample_weatherdata", List.of("data")
    );

    /**
     * Cache for sample data availability to avoid repeated database queries
     */
    private static final Map<String, Boolean> sampleDataCache = new ConcurrentHashMap<>();

    /**
     * Global flag to track if we've shown sample data availability summary
     */
    private static volatile boolean hasShownSummary = false;
    private static final Object summaryLock = new Object();

    private static String overrideConnectionString = null;

    /**
     * For testing: override the connection string used by the checker.
     * Pass null to clear the override and use the environment variable.
     */
    public static void setConnectionStringOverride(String uri) {
        overrideConnectionString = uri;
    }

    /**
     * Clears the sample data availability cache.
     * Useful for testing or when sample data availability may have changed.
     */
    public static void clearSampleDataCache() {
        sampleDataCache.clear();
        hasShownSummary = false;
    }

    /**
     * Creates a MongoDB client with appropriate timeouts for testing.
     * The caller is responsible for closing the returned client.
     */
    private static MongoClient createClientWithTimeouts(String connectionString) {
        MongoClientSettings settings = MongoClientSettings.builder()
            .applyConnectionString(new ConnectionString(connectionString))
            .applyToClusterSettings(builder ->
                builder.serverSelectionTimeout(2000L, TimeUnit.MILLISECONDS))
            .applyToSocketSettings(builder ->
                builder.connectTimeout(2000L, TimeUnit.MILLISECONDS)
                      .readTimeout(2000, TimeUnit.MILLISECONDS))
            .build();
        return MongoClients.create(settings);
    }

    private static String loadConnectionString() {
        // Try override first
        if (overrideConnectionString != null) {
            if (overrideConnectionString.isEmpty()) {
                return null;
            }
            return overrideConnectionString;
        }

        // Fall back to system environment
        return System.getenv(DEFAULT_CONN_STRING_ENV);
    }

    /**
     * Shows a helpful summary of sample data availability (only once per test run).
     * Returns immediately if already shown to avoid duplicate logging.
     */
    public static void showSampleDataSummary() {
        if (hasShownSummary) return;

        synchronized (summaryLock) {
            if (hasShownSummary) return;
            hasShownSummary = true;
        }

        try {
            List<String> availableDatabases = getAvailableSampleDatabases();

            if (availableDatabases.isEmpty()) {
                System.out.println("\n📊 Sample Data Status: No MongoDB sample databases found");
                System.out.println("   Some tests may be skipped. To load sample data:");
                System.out.println("   • Atlas: https://www.mongodb.com/docs/atlas/sample-data/");
                System.out.println("   • Local: Use mongorestore with sample data archive\n");
            } else {
                System.out.println(String.format("\n📊 Sample Data Status: %d database(s) available", availableDatabases.size()));
                System.out.println(String.format("   Found: %s\n", String.join(", ", availableDatabases)));
            }
        } catch (Exception e) {
            // Silently fail - we don't want to break test runs over this
        }
    }

    /**
     * Checks if a specific sample database and its expected collections exist
     *
     * @param databaseName The sample database name to check
     * @param requiredCollections Optional list of specific collections to verify
     * @return True if the sample database and collections exist
     */
    public static boolean checkSampleDataAvailable(String databaseName, List<String> requiredCollections) {
        List<String> collectionsToCheck = requiredCollections != null
            ? requiredCollections
            : STANDARD_SAMPLE_DATABASES.get(databaseName);
        if (collectionsToCheck == null) {
            return false;
        }

        String collectionsKey = collectionsToCheck.stream().sorted().collect(Collectors.joining(","));
        String cacheKey = databaseName + ":" + collectionsKey;

        return sampleDataCache.computeIfAbsent(cacheKey, key -> {
            String uri = loadConnectionString();
            if (uri == null || uri.isEmpty()) {
                System.err.printf("⚠️  sample data check for \"%s\": CONNECTION_STRING is empty — check that .env is loaded%n", databaseName);
                return false;
            }
            try (MongoClient client = createClientWithTimeouts(uri)) {
                // Check if database exists and collections are accessible by listing
                // collections directly. This avoids requiring the cluster-level
                // listDatabases privilege that Atlas M0 users often lack.
                List<String> existingCollections = new ArrayList<>();
                client.getDatabase(databaseName).listCollectionNames().into(existingCollections);

                if (existingCollections.isEmpty()) {
                    System.err.printf("⚠️  sample data check for \"%s\": no collections found (database may not exist or user may lack listCollections privilege)%n", databaseName);
                    return false;
                }

                // Check collections if specified
                for (String collection : collectionsToCheck) {
                    if (!existingCollections.contains(collection)) {
                        System.err.printf("⚠️  sample data check for \"%s\": missing collection \"%s\" (found: %s)%n", databaseName, collection, existingCollections);
                        return false;
                    }
                }

                return true;
            } catch (Exception e) {
                System.err.printf("⚠️  sample data check for \"%s\": %s%n", databaseName, e.getMessage());
                return false;
            }
        });
    }

    /**
     * Checks if a specific sample database and its expected collections exist
     *
     * @param databaseName The sample database name to check
     * @return True if the sample database exists with default collections
     */
    public static boolean checkSampleDataAvailable(String databaseName) {
        return checkSampleDataAvailable(databaseName, null);
    }

    /**
     * Checks if any of the standard MongoDB sample databases are available
     *
     * @return List of available sample database names
     */
    public static List<String> getAvailableSampleDatabases() {
        List<CompletableFuture<String>> futures = STANDARD_SAMPLE_DATABASES.keySet().stream()
            .map(dbName -> CompletableFuture.supplyAsync(() -> {
                if (checkSampleDataAvailable(dbName)) {
                    return dbName;
                }
                return null;
            }))
            .collect(Collectors.toList());

        return futures.stream()
            .map(CompletableFuture::join)
            .filter(Objects::nonNull)
            .collect(Collectors.toList());
    }

    /**
     * Checks if multiple sample databases are available
     *
     * @param requiredDatabases The sample database names to check
     * @param collectionsPerDatabase Optional map of database names to required collections
     * @return SampleDataAvailability result containing availability status and missing databases
     */
    public static SampleDataAvailability checkMultipleSampleDatabases(
            List<String> requiredDatabases,
            Map<String, List<String>> collectionsPerDatabase) {

        List<CompletableFuture<DatabaseCheckResult>> futures = requiredDatabases.stream()
            .map(dbName -> CompletableFuture.supplyAsync(() -> {
                List<String> requiredCollections = collectionsPerDatabase != null
                    ? collectionsPerDatabase.get(dbName) : null;
                boolean isAvailable = checkSampleDataAvailable(dbName, requiredCollections);
                return new DatabaseCheckResult(dbName, isAvailable);
            }))
            .collect(Collectors.toList());

        List<DatabaseCheckResult> results = futures.stream()
            .map(CompletableFuture::join)
            .collect(Collectors.toList());

        List<String> missingDatabases = results.stream()
            .filter(r -> !r.isAvailable())
            .map(DatabaseCheckResult::databaseName)
            .collect(Collectors.toList());

        List<String> availableDatabases = results.stream()
            .filter(DatabaseCheckResult::isAvailable)
            .map(DatabaseCheckResult::databaseName)
            .collect(Collectors.toList());

        return new SampleDataAvailability(
            missingDatabases.isEmpty(),
            missingDatabases,
            availableDatabases
        );
    }

    /**
     * Checks if multiple sample databases are available
     *
     * @param requiredDatabases The sample database names to check
     * @return SampleDataAvailability result containing availability status and missing databases
     */
    public static SampleDataAvailability checkMultipleSampleDatabases(List<String> requiredDatabases) {
        return checkMultipleSampleDatabases(requiredDatabases, null);
    }

    /**
     * Internal record for database check results
     */
    private record DatabaseCheckResult(String databaseName, boolean isAvailable) {}
}
